File: util/src/main/java/cn/readsense/module/base/BaseCoreActivity.java
package cn.readsense.module.base;

import android.app.ProgressDialog;
import android.content.Context;
import android.os.Bundle;

import androidx.annotation.IdRes;
import androidx.annotation.LayoutRes;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.lifecycle.LifecycleObserver;

import java.util.ArrayList;
import java.util.List;

import butterknife.ButterKnife;
import butterknife.Unbinder;
import cn.readsense.module.permissions.PermissionListener;
import cn.readsense.module.permissions.PermissionsUtil;
import cn.readsense.module.util.DLog;
import cn.readsense.module.util.DisplayUtil;
import cn.readsense.module.util.ToastUtils;


public abstract class BaseCoreActivity extends AppCompatActivity {
    public ProgressDialog progressDialog;
    Unbinder unbinder;
    public Context context;

    int screenWidth, screenHeight;
    private String permissions[];
    private List<LifecycleObserver> lifecycleObservers = new ArrayList<>();

    @Override
    public void setContentView(@LayoutRes int layoutResID) {
        super.setContentView(layoutResID);
        unbinder = ButterKnife.bind(this);
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        context = this;
        progressDialog = new ProgressDialog(context);
        progressDialog.setCancelable(false);
        progressDialog.setCanceledOnTouchOutside(false);

        screenWidth = DisplayUtil.getScreenWidth(context);
        screenHeight = DisplayUtil.getScreenHeight(context);
        final int layoutId = getLayoutId();
        if (layoutId != 0) {
            setContentView(layoutId);

            if (permissions != null) {
                if (!PermissionsUtil.hasPermission(context, permissions)) {
                    PermissionsUtil.requestPermission(context, new PermissionListener() {
                        @Override
                        public void permissionGranted(@NonNull String[] permission) {
                            registerViewAndObserve();
                        }

                        @Override
                        public void permissionDenied(@NonNull String[] permission) {
                            ToastUtils.show("相关请同意权限！！");
                        }
                    }, permissions);
                } else {
                    registerViewAndObserve();
                }
            } else {
                registerViewAndObserve();
            }
        } else {
            throw new Error("internal error: layoutId = 0!!");
        }
    }

    private void registerViewAndObserve() {
        initView();
        if (lifecycleObservers != null && lifecycleObservers.size() > 0) {
            for (LifecycleObserver observer : lifecycleObservers) {
                getLifecycle().addObserver(observer);
            }
        }
    }


    public void requestPermissions(String... permissions) {
        this.permissions = permissions;
    }

    public void addLifecycleObserver(LifecycleObserver lifecycleObserver) {
        if (!lifecycleObservers.contains(lifecycleObserver)) {
            lifecycleObservers.add(lifecycleObserver);
        } else {
            throw new Error("internal error: observer already input!!");
        }
    }

    public void showToast(String msg) {
        ToastUtils.show(msg);
    }

    public void showProgress(String msg) {
        progressDialog.setMessage(msg);
        progressDialog.show();
    }

    public void dismissProgress() {
        if (progressDialog != null) {
            progressDialog.dismiss();
        }
    }

    public int getScreenWidth() {
        return screenWidth;
    }

    public int getScreenHeight() {
        return screenHeight;
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        unbinder.unbind();
    }

    protected abstract int getLayoutId();

    protected abstract void initView();


    /**
     * 添加fragment
     *
     * @param fragment
     * @param frameId
     */
    protected void addFragment(BaseCoreFragment fragment, @IdRes int frameId) {
        Utils.checkNotNull(fragment);
        getSupportFragmentManager().beginTransaction()
                .add(frameId, fragment, fragment.getClass().getSimpleName())
                .addToBackStack(fragment.getClass().getSimpleName())
                .commitAllowingStateLoss();

    }


    /**
     * 替换fragment
     *
     * @param fragment
     * @param frameId
     */
    protected void replaceFragment(BaseCoreFragment fragment, @IdRes int frameId) {
        Utils.checkNotNull(fragment);
        getSupportFragmentManager().beginTransaction()
                .replace(frameId, fragment, fragment.getClass().getSimpleName())
                .addToBackStack(fragment.getClass().getSimpleName())
                .commitAllowingStateLoss();

    }


    /**
     * 隐藏fragment
     *
     * @param fragment
     */
    protected void hideFragment(BaseCoreFragment fragment) {
        Utils.checkNotNull(fragment);
        getSupportFragmentManager().beginTransaction()
                .hide(fragment)
                .commitAllowingStateLoss();

    }


    /**
     * 显示fragment
     *
     * @param fragment
     */
    protected void showFragment(BaseCoreFragment fragment) {
        Utils.checkNotNull(fragment);
        getSupportFragmentManager().beginTransaction()
                .show(fragment)
                .commitAllowingStateLoss();

    }


    /**
     * 移除fragment
     *
     * @param fragment
     */
    protected void removeFragment(BaseCoreFragment fragment) {
        Utils.checkNotNull(fragment);
        getSupportFragmentManager().beginTransaction()
                .remove(fragment)
                .commitAllowingStateLoss();

    }


    /**
     * 弹出栈顶部的Fragment
     */
    protected void popFragment() {
        if (getSupportFragmentManager().getBackStackEntryCount() > 1) {
            getSupportFragmentManager().popBackStack();
        } else {
            finish();
        }
    }
}

File: util/src/main/java/cn/readsense/module/base/BaseCoreActivity.kt
package cn.readsense.module.base

import android.app.ProgressDialog
import android.os.Bundle
import androidx.annotation.IdRes
import androidx.annotation.LayoutRes
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.LifecycleObserver
import cn.readsense.module.permissions.PermissionListener
import cn.readsense.module.permissions.PermissionsUtil
import cn.readsense.module.util.DisplayUtil
import cn.readsense.module.util.ToastUtils
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.MainScope
import kotlinx.coroutines.cancel

abstract class BaseCoreActivity : AppCompatActivity(), CoroutineScope by MainScope() {
    var progressDialog: ProgressDialog? = null
    private var screenWidth = 0
    private var screenHeight = 0
    private var permissions: MutableList<String> = mutableListOf()
    private val lifecycleObservers: MutableList<LifecycleObserver> = mutableListOf()
    override fun setContentView(@LayoutRes layoutResID: Int) {
        super.setContentView(layoutResID)
    }

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        progressDialog = ProgressDialog(baseContext)
        progressDialog!!.setCancelable(false)
        progressDialog!!.setCanceledOnTouchOutside(false)
        screenWidth = DisplayUtil.getScreenWidth(baseContext)
        screenHeight = DisplayUtil.getScreenHeight(baseContext)
        val layoutId = getLayoutId()
        if (layoutId != 0) {
            setContentView(layoutId)
            if (!PermissionsUtil.hasPermission(baseContext, *permissions.toTypedArray())) {
                PermissionsUtil.requestPermission(baseContext, object : PermissionListener {
                    override fun permissionGranted(permission: Array<String>) {
                        registerViewAndObserve()
                    }

                    override fun permissionDenied(permission: Array<String>) {
                        ToastUtils.show("相关请同意权限！！")
                    }
                }, *permissions.toTypedArray())
            } else {
                registerViewAndObserve()
            }
        } else {
            throw Error("internal error: layoutId = 0!!")
        }
    }

    private fun registerViewAndObserve() {
        initView()
        if (lifecycleObservers.size > 0) {
            for (observer in lifecycleObservers) {
                lifecycle.addObserver(observer)
            }
        }
    }

    fun requestPermissions(vararg permissions: String) {
        this.permissions.clear()
        for (permission in permissions)
            this.permissions.add(permission)

    }

    fun addLifecycleObserver(lifecycleObserver: LifecycleObserver) {
        if (!lifecycleObservers!!.contains(lifecycleObserver)) {
            lifecycleObservers.add(lifecycleObserver)
        } else {
            throw Error("internal error: observer already input!!")
        }
    }

    fun showToast(msg: String?) {
        ToastUtils.show(msg)
    }

    fun showProgress(msg: String?) {
        progressDialog!!.setMessage(msg)
        progressDialog!!.show()
    }

    fun dismissProgress() {
        if (progressDialog != null) {
            progressDialog!!.dismiss()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        cancel()
    }

    protected abstract fun getLayoutId(): Int

    protected abstract fun initView()

    /**
     * 添加fragment
     *
     * @param fragment
     * @param frameId
     */
    protected fun addFragment(fragment: BaseCoreFragment, @IdRes frameId: Int) {
        Utils.checkNotNull(fragment)
        supportFragmentManager.beginTransaction()
            .add(frameId, fragment, fragment.javaClass.simpleName)
            .addToBackStack(fragment.javaClass.simpleName)
            .commitAllowingStateLoss()
    }

    /**
     * 替换fragment
     *
     * @param fragment
     * @param frameId
     */
    protected fun replaceFragment(fragment: BaseCoreFragment, @IdRes frameId: Int) {
        Utils.checkNotNull(fragment)
        supportFragmentManager.beginTransaction()
            .replace(frameId, fragment, fragment.javaClass.simpleName)
            .addToBackStack(fragment.javaClass.simpleName)
            .commitAllowingStateLoss()
    }

    /**
     * 隐藏fragment
     *
     * @param fragment
     */
    protected fun hideFragment(fragment: BaseCoreFragment) {
        Utils.checkNotNull(fragment)
        supportFragmentManager.beginTransaction()
            .hide(fragment)
            .commitAllowingStateLoss()
    }

    /**
     * 显示fragment
     *
     * @param fragment
     */
    protected fun showFragment(fragment: BaseCoreFragment) {
        Utils.checkNotNull(fragment)
        supportFragmentManager.beginTransaction()
            .show(fragment)
            .commitAllowingStateLoss()
    }

    /**
     * 移除fragment
     *
     * @param fragment
     */
    protected fun removeFragment(fragment: BaseCoreFragment) {
        Utils.checkNotNull(fragment)
        supportFragmentManager.beginTransaction()
            .remove(fragment)
            .commitAllowingStateLoss()
    }

    /**
     * 弹出栈顶部的Fragment
     */
    protected fun popFragment() {
        if (supportFragmentManager.backStackEntryCount > 1) {
            supportFragmentManager.popBackStack()
        } else {
            finish()
        }
    }
}